### ISSUE TITLE
Difference in reliability between QUIC vs UDP

### ISSUE THREAD

**Issue by godmodegalactus** [solana-labs/solana#28549]
#### Problem

We have made a project to benchmark mango-v3 on solana. We tested our benchmark with UDP vs QUIC and we found out using TPUClient in UDP mode on low tps all our transactions are confirmed but in QUIC mode only 25% are confirmed and others are lost (timeouts). 75% of transaction are never added in the transaction queue of the TPU.

Please check the following project for reference : 
`https://github.com/godmodegalactus/mango_bencher`

Let me know if you have issues running the projects.
I also noted that I get lot of errors which I did not had in UDP mode (these exceptions are defined in quinn repo).

```
[2022-10-19T19:10:29.814314095Z WARN  solana_metrics::metrics] datapoint: send-wire-async failure=1i
[2022-10-19T19:10:29.814338228Z WARN  solana_metrics::metrics] datapoint: send-wire-async failure=1i
[2022-10-19T19:10:29.814265901Z INFO  solana_client::nonblocking::quic_client] Error sending to 127.0.0.1:1033 with id 140704337396432, error WriteError(ConnectionLost(ApplicationClosed(ApplicationClose { error_code: 0, reason: b"" }))) thread: ThreadId(36)
[2022-10-19T19:10:29.814281278Z INFO  solana_client::nonblocking::quic_client] Error sending to 127.0.0.1:1033 with id 140704337396432, error WriteError(ConnectionLost(ApplicationClosed(ApplicationClose { error_code: 0, reason: b"" }))) thread: ThreadId(48)
[2022-10-19T19:10:29.814363160Z WARN  solana_metrics::metrics] datapoint: send-wire-async failure=1i
[2022-10-19T19:10:29.814477980Z INFO  solana_client::nonblocking::quic_client] Error sending to 127.0.0.1:1033 with id 140704337396432, error WriteError(ConnectionLost(ApplicationClosed(ApplicationClose { error_code: 0, reason: b"" }))) thread: ThreadId(49)
[2022-10-19T19:10:29.814536551Z WARN  solana_client::nonblocking::quic_client] Failed to send transaction async to 127.0.0.1:1033, error: Custom("WriteError(ConnectionLost(ApplicationClosed(ApplicationClose { error_code: 0, reason: b\"\" })))") 
```

I have tested this on testnet and localnet.

We process blocks to confirm transactions and create metrics and I noticed that in UDP mode all blocks have multiple transactions which were sent by the benchmarker but in case of QUIC there are spurs where multiple blocks which do not contain any of these transactions. This was also tested on localnet which was used only by the benchmarker.

last few logs for processing blocks of mango bencher in udp mode : 
```
[2022-10-22T20:55:52.624612963Z INFO  solana_bench_mango] block 7Nt5JNmdTmQQNmqKyioLFmW7HucqCFTPGQrRKQ6NFF1G at slot 5068 contains 271 transactions
[2022-10-22T20:55:52.911545873Z INFO  solana_bench_mango] block 4Qwt4xHPuAfpRCLmpSB1ojGFT8M6iJeAR7czbbzGL5ya at slot 5086 contains 403 transactions
[2022-10-22T20:55:52.976742789Z INFO  solana_bench_mango] block 6sr22B51y6sfU3S3x7tcudvdovL8KhxGDwhBFWTugv4K at slot 5074 contains 355 transactions
[2022-10-22T20:55:53.022407800Z INFO  solana_bench_mango] block FYeuFXWx1B3RmrH1ytmYHZDdSGZ8LWLsZ9Admtu85sxK at slot 5080 contains 359 transactions
[2022-10-22T20:55:53.040483276Z INFO  solana_bench_mango] block HVgZhJkfyXbEDJWEZHc8BvcgpBYMFLrGqXuVSjw5h3UA at slot 5069 contains 237 transactions
[2022-10-22T20:55:53.461355070Z INFO  solana_bench_mango] block 693aHZcnzBskDtChL4LYZt9X4y5jhFT3dmMZSaaFJ34j at slot 5087 contains 370 transactions
[2022-10-22T20:55:53.679178177Z INFO  solana_bench_mango] block 5gDReEY4LRphNRXPGdW9QxdA96MP4pVuNiKF6tAgBJxV at slot 5075 contains 356 transactions
[2022-10-22T20:55:53.794391787Z INFO  solana_bench_mango] block AMZ128AHqhPg5E1E3pEswAUSMuehGS8u5wiTzRGFtTNT at slot 5070 contains 354 transactions
[2022-10-22T20:55:53.799966004Z INFO  solana_bench_mango] block 5amccpZNYy1gdwo3ZmTz2adVJkuPgeZZWz1vPY6So12V at slot 5081 contains 385 transactions
[2022-10-22T20:55:53.951789301Z INFO  solana_bench_mango] block 4oQzgvWZWfnAv3NMyhpThxSmE1VxEsGNCcW5CCzPpYXm at slot 5088 contains 349 transactions
[2022-10-22T20:55:54.079220363Z INFO  solana_bench_mango] block B5kCChkJ8MsUs8d9zF2NmkU3vT9msYSj6dQTMo6jKjjJ at slot 5071 contains 173 transactions
[2022-10-22T20:55:54.157196503Z INFO  solana_bench_mango] block GnYAAEyqrC8YWfWhB7k1aE6YXyqg8nkn8anGuR9pUM4x at slot 5076 contains 343 transactions
[2022-10-22T20:55:54.320836846Z INFO  solana_bench_mango] block 1451JCEnkFHHA9RqHdtVMM4pbaUqK3Wnme5GeHs9h5o5 at slot 5082 contains 369 transactions
[2022-10-22T20:55:54.555285521Z INFO  solana_bench_mango] block 3vXxtBZemfd5hd5TxZ58xRAvvnLjZTWgaRZ7W8oKo7dx at slot 5089 contains 395 transactions
[2022-10-22T20:55:54.742579858Z INFO  solana_bench_mango] block 9BGg5EbZvpxyiWrfR49X1C892eNsi2Z6f2aFuECdPTiP at slot 5072 contains 343 transactions
[2022-10-22T20:55:54.783607422Z INFO  solana_bench_mango] block CewNQQERhE6cTm6TqD1bdQZqoVdHnzSnjzhYXJcSMyuF at slot 5077 contains 343 transactions
[2022-10-22T20:55:54.924229627Z INFO  solana_bench_mango] block 5zenx3KqKPyDA6n1f7GNufXw9n8yzFhcx9b2WYJesNoh at slot 5083 contains 393 transactions
[2022-10-22T20:55:55.054020974Z INFO  solana_bench_mango] block F3Wuohngv9zDvS9iUwbKMxqe1PFfH9XutY8d6C4f5TLd at slot 5090 contains 400 transactions
[2022-10-22T20:55:55.348547403Z INFO  solana_bench_mango] block 91AcpanUa1zTxiveJyboR53TWHSvZLW2SMw4gYi6hzn at slot 5073 contains 383 transactions
[2022-10-22T20:55:55.398600072Z INFO  solana_bench_mango] block HXW52XCDEC6Y37kmhw6x2g2gkXKoD2dmqxjQWL5Y5Qy2 at slot 5078 contains 357 transactions
[2022-10-22T20:55:55.544342126Z INFO  solana_bench_mango] block 9CrkvAfaavYKTZEhEsjazadSCkce42jJEhtzv2nhJBPy at slot 5084 contains 416 transactions
[2022-10-22T20:55:55.595421163Z INFO  solana_bench_mango] block HwiCcf58GzDzXykddCVdQXuKuhNLjGq7PzmCLnC41C44 at slot 5091 contains 412 transactions
[2022-10-22T20:55:55.791413663Z INFO  solana_bench_mango] block FBEkc1MsW1EyzMjEyDpT8jQoKXAqemzvTbp6xY8w6cGR at slot 5079 contains 380 transactions
[2022-10-22T20:55:55.940606295Z INFO  solana_bench_mango] block 35Fn81yTJJqVQaJ7EHrotiN4CbWQVWpmeRmSUZD3ucKQ at slot 5085 contains 382 transactions
```

And for QUIC : 

```
[2022-10-23T07:50:20.480402882Z INFO  solana_bench_mango] block CoJEKmnk1xj8uobiDw1E7KC3kYWu9peotixWMW6y35b1 at slot 3897 contains 3 transactions and consumerd 70967 CUs
[2022-10-23T07:50:20.480515956Z INFO  solana_bench_mango] block BmHnTXcZGD6id3dbKFNwtTnU6aVkBRPWaYYbZLgAf4Kx at slot 3898 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.480546194Z INFO  solana_bench_mango] block A78mSZHP7uVSMCxazSNVJ944Jwu99Mbn8VRnPSdsHbCJ at slot 3899 contains 223 transactions and consumerd 24511435 CUs
[2022-10-23T07:50:20.480574396Z INFO  solana_bench_mango] block 5UNZcZXyyu75jvXptZCMuk5x6jNSanQVWgpVtemuHh64 at slot 3900 contains 103 transactions and consumerd 11263260 CUs
[2022-10-23T07:50:20.480601812Z INFO  solana_bench_mango] block 72SZ6yu5Ay5PPbGtRhSZ3n3F9sgqsbiuB3jvGU3z5fHz at slot 3901 contains 211 transactions and consumerd 23252002 CUs
[2022-10-23T07:50:20.480629132Z INFO  solana_bench_mango] block AvjkniC8ucLV2N3GKCWmWbcqiSN6xCjiqgiKKzpYpggq at slot 3902 contains 75 transactions and consumerd 8001516 CUs
[2022-10-23T07:50:20.480656700Z INFO  solana_bench_mango] block 7Dp3oKVw6MJSydV6H53CNzKg3k3hGHe3WhEQ9uqrcvEw at slot 3903 contains 180 transactions and consumerd 19712167 CUs
[2022-10-23T07:50:20.480684340Z INFO  solana_bench_mango] block BQHfqTAi7ZzAdNHj2V9UHSfbVDvKtLEZWutJiwBwsxpG at slot 3904 contains 131 transactions and consumerd 14128943 CUs
[2022-10-23T07:50:20.480711695Z INFO  solana_bench_mango] block DrXZHxZaphLE43SLVBrWcFYe1p3vmhdi6vzStFrEFwsc at slot 3905 contains 160 transactions and consumerd 17277482 CUs
[2022-10-23T07:50:20.480739139Z INFO  solana_bench_mango] block EiK2hBPpxpNK4Qi5Cs3Aqb6PcVGdR1H4zWqkRuW8tjN4 at slot 3906 contains 151 transactions and consumerd 16549861 CUs
[2022-10-23T07:50:20.480766428Z INFO  solana_bench_mango] block 3BxUfUnXsBYTwEYNgZUyYskK7TzmLqG9jEd1hbW27gUJ at slot 3907 contains 157 transactions and consumerd 17044433 CUs
[2022-10-23T07:50:20.480793101Z INFO  solana_bench_mango] block A9jrPqrFutTMU3uQ5ck8ZXdRMdTN8C2JbNj1xNDtYndf at slot 3908 contains 157 transactions and consumerd 17102241 CUs
[2022-10-23T07:50:20.480820185Z INFO  solana_bench_mango] block 8oSDuqz663xrhoFoFwk18eFEH179Unny9hSqp3D12Ttc at slot 3909 contains 154 transactions and consumerd 16808692 CUs
[2022-10-23T07:50:20.480847908Z INFO  solana_bench_mango] block BKBEg1Hri3TGYePk2J53kkopXt1eQNry6tLXvDDrfrhw at slot 3910 contains 162 transactions and consumerd 17484644 CUs
[2022-10-23T07:50:20.480874997Z INFO  solana_bench_mango] block xwMtaLs3oNseyuw26LZieCovgM7joaeqLXfJQwWPUfR at slot 3911 contains 159 transactions and consumerd 17491779 CUs
[2022-10-23T07:50:20.480901520Z INFO  solana_bench_mango] block GQZbVsPTd9sxooUH6htyvrrAMeeHcuYQYaPDaJHAtAp9 at slot 3912 contains 155 transactions and consumerd 16686127 CUs
[2022-10-23T07:50:20.480928769Z INFO  solana_bench_mango] block 278yUtgKb1v1hPk1d5gJh76WCU9n6dggtFkKYTM4qwf5 at slot 3913 contains 139 transactions and consumerd 15263999 CUs
[2022-10-23T07:50:20.480956122Z INFO  solana_bench_mango] block 5uN2jv3pdP2mdpd1mrYw2rkqJMCRHuWW8bxiXWef577J at slot 3914 contains 170 transactions and consumerd 18548406 CUs
[2022-10-23T07:50:20.480983612Z INFO  solana_bench_mango] block 5des75N2rJ7F3N8RK8co5P2MtX4DnPjttHrL8YUa3Hut at slot 3915 contains 111 transactions and consumerd 12025346 CUs
[2022-10-23T07:50:20.481010701Z INFO  solana_bench_mango] block E1EXV7JdsprF7rwEYw4Nima5R2JEyPoEsRtyCf2FFy9t at slot 3916 contains 202 transactions and consumerd 22100448 CUs
[2022-10-23T07:50:20.481038121Z INFO  solana_bench_mango] block A4yVCcwJ19Au3BPYAn5NbHf4pTRrRWAD142dz7otuF87 at slot 3917 contains 78 transactions and consumerd 8257426 CUs
[2022-10-23T07:50:20.481065284Z INFO  solana_bench_mango] block 4E9ssd61HPuGrAoQjCPSRixWj6kS5mdv2S2RneRPrSno at slot 3918 contains 239 transactions and consumerd 26157048 CUs
[2022-10-23T07:50:20.481092445Z INFO  solana_bench_mango] block 6Vj2ncn3tyv1EenzLuVa71gPMwMX13jdJZLTFQm86S3z at slot 3919 contains 10 transactions and consumerd 740399 CUs
[2022-10-23T07:50:20.481119616Z INFO  solana_bench_mango] block 8cLEMtTb3PXZjg6ZcXmUrrhXkiqWSs2yBiZvb425DpJY at slot 3920 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481146224Z INFO  solana_bench_mango] block 34XZ8Ep7LG72tVpZxvJtS5zm7k694xws9VjCGwbjHBtb at slot 3921 contains 11 transactions and consumerd 642307 CUs
[2022-10-23T07:50:20.481173143Z INFO  solana_bench_mango] block 8JpjF54C467GKnNCZLjLKEJNBRvo9KeRHZzFAnk6GkWT at slot 3922 contains 3 transactions and consumerd 154369 CUs
[2022-10-23T07:50:20.481200409Z INFO  solana_bench_mango] block 8yspjke41XJAycVQLMGeDRmdJoT3HecDGKDY3RB2iWze at slot 3923 contains 3 transactions and consumerd 103958 CUs
[2022-10-23T07:50:20.481227274Z INFO  solana_bench_mango] block 6FVnYAto9APVDb63qiKSSLFnU2pcy2UJLmbTKnfGozbx at slot 3924 contains 8 transactions and consumerd 317189 CUs
[2022-10-23T07:50:20.481253755Z INFO  solana_bench_mango] block BBDm3KRt43XBoJBmXHuutKPN7s8Sq8KxXJp8Cqm5QDnp at slot 3925 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481280144Z INFO  solana_bench_mango] block BDG1sx32TwHJihaDdMhFsAWADDb4wjUc1ffn6QUab8Qm at slot 3926 contains 3 transactions and consumerd 92093 CUs
[2022-10-23T07:50:20.481306703Z INFO  solana_bench_mango] block GxtZdPj4svPH2Kwk1SW3MCeGCYT6ZHTnuhnBBW6LetSJ at slot 3927 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481332897Z INFO  solana_bench_mango] block FuGfd8299KtjXMmWiFyLohqTfPAbw94q67mFsqMDsd4o at slot 3928 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481359137Z INFO  solana_bench_mango] block 41PcDVxNzwaxF7BSpDLVsVDsZGu7X6Y94E1cQBY6xkuu at slot 3929 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481386096Z INFO  solana_bench_mango] block 7u2ZYgLz9n8WwnYKe7HcJhsiMC8M2RFhyTTpx5KariTF at slot 3930 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481412847Z INFO  solana_bench_mango] block 6Cep8yWgtGjctZS9hyo67TL99zFKVd3j64TP698pYDAc at slot 3931 contains 3 transactions and consumerd 87310 CUs
[2022-10-23T07:50:20.481440152Z INFO  solana_bench_mango] block 5KoHo1Tdmf6JywFXn7oyqExDTnTqGQzqJ43h8brVP8D1 at slot 3932 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481466719Z INFO  solana_bench_mango] block 67ghzA15gJadn8DqYAY4y2CnpSKvVFifsJBRWs5rWAuv at slot 3933 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481493090Z INFO  solana_bench_mango] block HbX4BviZUFGeFJJsNSrqNR9Ma1MGpAMPUrZqVCTrYrSs at slot 3934 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481519711Z INFO  solana_bench_mango] block 4XaGpc9wSVybhLup1Q3HqKkP7dmTFJsyCgARBVSbwT96 at slot 3935 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481545967Z INFO  solana_bench_mango] block CGQUGcYiAaFLbZ4pJ4dt4RRQtpeNs7PUkPy4uH9uzZX6 at slot 3936 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481573101Z INFO  solana_bench_mango] block ERrpEhNCKSE87cTj34WwFtbNbp3k1QTo8gxufCery2Ld at slot 3937 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481599149Z INFO  solana_bench_mango] block 2hg9MjCfM1dBJUrQL2MG1twZGUYGCzxfu4PXT7BHM2HQ at slot 3938 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481625549Z INFO  solana_bench_mango] block HLVu9ny2QTSsDrcxJSCrLmTpVGYhwnssYhRyQPf8DrUb at slot 3939 contains 3 transactions and consumerd 70967 CUs
[2022-10-23T07:50:20.481652716Z INFO  solana_bench_mango] block CpBsrJTNw6siWaV5JBSQXYLA3sJKWPgvHsGn7pKqHvcf at slot 3940 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481679129Z INFO  solana_bench_mango] block E6rrPitoCmDjDGHV6u9i81jVM5uofaqyrHJrTgiCQtYj at slot 3941 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481705368Z INFO  solana_bench_mango] block A8PXujnjZjsfWoHvn4MSDtCKRS5WD1qdrqaaRaDJEKf3 at slot 3942 contains 5 transactions and consumerd 313028 CUs
[2022-10-23T07:50:20.481731859Z INFO  solana_bench_mango] block GCFGzB6rz1VsBP7cuMYLnBiVwhcpN2cszwEUB4pneXYj at slot 3943 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481758164Z INFO  solana_bench_mango] block pcUP8eLLtgTQKqzgW8FdmKDg9n6wivff174YcdXdnpk at slot 3944 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481784704Z INFO  solana_bench_mango] block GYZL7tNxX6Qr2hJumpN6SV39aZWDfHaXJDbS6UyTi6d3 at slot 3945 contains 1 transactions and consumerd 0 CUs
[2022-10-23T07:50:20.481811093Z INFO  solana_bench_mango] block 9GDNPSFMwLAkyrAtMhaA7TzJ4p2nTiGBBMRm9qdUCXfF at slot 3946 contains 3 transactions and consumerd 87310 CUs

```

#### Proposed Solution
Find out and fix why does this happen.

**Comment by KirillLykov** [comment]
Very interesting, thanks for sharing! I will check this out in the next couple of weeks if it is fine

**Comment by KirillLykov** [comment]
@godmodegalactus could you share instructions how to run tool on localnet?

**Comment by godmodegalactus** [comment]
Hey there is a project configure mango here: https://github.com/godmodegalactus/configure_mango

This project will configure the local cluster and give as output all the necessary files.

Check the readme file. Let me know if something is unclear. Should be pretty smooth.

**Comment by KirillLykov** [comment]
@godmodegalactus I didn't manage to launch mango using provided scripts, see https://github.com/godmodegalactus/configure_mango/issues/1

**Comment by KirillLykov** [comment]
To document the way mango_bench was launched:
```
cd configure_mango
./configure_local.sh # for validator
npm install
yarn add @blockworks-foundation/mango-client
# replace in index.ts localhost with 0.0.0.0 if solana v 1.15 is used
ts-node index.ts # creates ids.json and other files
ts-node keeper.ts # to update mango price cache to avoid getting mango errors about stale price
cd ../bench-mango
cargo run --bin solana-bench-mango -- -u http://0.0.0.0:8899 --identity ../configure_mango/authority.json --accounts ../configure_mango/accounts-20.json  --mango ../configure_mango/ids.json --mango-cluster localnet --duration 10 -q 2 --transaction_save_file tlog.csv --block_data_save_file blog.csv 2> err.txt
```

**Comment by KirillLykov** [comment]
Mostly resolved by a series of patches to mango_bencher along with sticking with the same solana version as used on testnet.
Remaining issue is that we observe quite some errors 
```bash
2022-12-12T21:04:55.973966860Z INFO  solana_client::quic_client] Timedout sending transaction 141.94.97.81:8009
```

The reason is unclear to me.

**Comment by godmodegalactus** [comment]
Yes i agree the performance of QUIC is better than before.

**Comment by bardo-hub** [comment]
m

**Comment by KirillLykov** [comment]
One of the difference between UDP and QUIC on our side is that in the later batches are formed of 1 transaction always while with UDP it could be 128. See https://github.com/solana-labs/solana/blob/1c862f0b666c6de251a630d9128313b51dfa7c46/streamer/src/nonblocking/quic.rs#L670

**Comment by lijunwangs** [comment]
> Yes i agree the performance of QUIC is better than before.

@godmodegalactus Is this an issue with the latest code in master or v1.13, v1.14?

**Comment by godmodegalactus** [comment]
> > Yes i agree the performance of QUIC is better than before.
> 
> @godmodegalactus Is this an issue with the latest code in master or v1.13, v1.14?

The results are much better than before but I still have lot of transactions getting dropped due to https://github.com/solana-labs/solana/issues/29895 .
Kirill and me are also looking into testnet results.

**Comment by lijunwangs** [comment]
Can this issue be resolved then? https://github.com/solana-labs/solana/issues/29895 seems to be an issue at the bank stage? Or are you saying it happens more easily because we used quic?

**Comment by godmodegalactus** [comment]
> Can this issue be resolved then? #29895 seems to be an issue at the bank stage? Or are you saying it happens more easily because we used quic?

@KirillLykov If you have finished your analysis its your call. I think it is better than before, If i feel that QUIC still adds more issues we can reopen this issue.

**Comment by KirillLykov** [comment]
> Can this issue be resolved then? #29895 seems to be an issue at the bank stage? Or are you saying it happens more easily because we used quic?

If I understand correctly the issue is about reduction of the TPS on QUIC vs UDP for unstaked clients. It is still present and for newer versions it is more sever because client in this case is unstaked. So on master the staked-based CC mechanism is more restrictive.
I think the answer is that it is done that way by design, not it is expected behavior. But for benchmarking on testnet we need to invent something to overcome CC. Proposed workaround is to use identity of a validator running on the same node). Atm we don't know if the workaround works.
